### PR TITLE
events.py: enhance efficiency of clear() by eliminating full dictionary rebuild

### PR DESCRIPTION
**Pull request by deanlee**
The current `clear() `method rebuilds the entire dictionary from the keys in `EVENTS` on each call, resulting in inefficient performance. For example, in `selfdrived,`  clear() is among the top five most time-consuming operations, with its execution time just behind  `SelfdriveD.publish_selfdriveState(self, CS)`, which is unreasonable:
```

     5000    0.192    0.000    0.467    0.000 selfdrived.py:415(publish_selfdriveState)
     5000    0.042    0.000    0.324    0.000 events.py:68(clear)
```
After this PR, the execution time of `clear()` has been reduced from 0.324  to 0.0311, making it approximately 10 times faster. This improvement should enhance the runtime efficiency of both selfdrived and card.

**Changes:**

1. Added self.prev_event_set to track previously active events.
2. Refactored clear() to use set operations for efficient counter reset instead of rebuilding the entire dictionary.
3. Updated add() method to increment event counters directly.

### CODE REVIEW

**Comment by deanlee** [comment]
Using defaultdict is more appropriate in the `Events` class. This is because Events() might be created and returned in various functions, such as create_common_events(). Initializing the counters  from the keys in EVENTS  in the __init__ method offers no real benefit other than wasting CPU time.  defaultdict simplifies the process and avoids unnecessary initialization.

File: selfdrive/selfdrived/events.py [hunk 1]
```diff
@@ -65,9 +67,20 @@ def add(self, event_name: int, static: bool=False) -> None:
       bisect.insort(self.static_events, event_name)
     bisect.insort(self.events, event_name)
 
+    self.event_counters[event_name] += 1
+
   def clear(self) -> None:
-    self.event_counters = {k: (v + 1 if k in self.events else 0) for k, v in self.event_counters.items()}
+     # Get events no longer active
```
Comment (sshane) [inline selfdrive/selfdrived/events.py:73]: This looks immediately more confusing to read, anything else we can try?

Comment (deanlee) [reply selfdrive/selfdrived/events.py:73]: A more common and readable approach is to loop through event_counters and check if event_name is not in the events list. The code snippet would look like this:
```
for event_name in self.event_counters:
    if event_name not in self.events:
        self.event_counters[event_name] = 0
```
However, this approach is significantly less efficient than the version in this PR and might not offer much improvement over the master version. I'll consider if there's a better solution.

Comment (deanlee) [reply selfdrive/selfdrived/events.py:73]: I can't think of a better solution for now. I've improved the comments to make the code easier to understand. The logic is straightforward: maintain a list of previously active events, compare it with the current active events, and remove those that are no longer active.

**Comment by adeebshihadeh** [comment]
trigger-jenkins

**Pull request closed without merging**